### PR TITLE
refactor full_node add_block

### PR DESCRIPTION
**Pull request by almogdepaz**
### Purpose:
full node add_block is only used to add blocks on top of the current Peak, it does not need ForkInfo besides from the usage in tests, this pr removes that field and updates all the tests, this allows us to remove cases from blockchain.py add_block that dont serve us in node operation 

this or also turns fork_info in full_node add_block_batch into a mandatory field this is possible since in normal sync flow we always now the fork before we call the add_block_batchs this allows removing code covering the fork_info == None case that is used in tests

### CODE REVIEW

**Review by arvidn — COMMENTED** [review]
looks really good!

**Comment by almogdepaz** [comment]
@arvid 
from what i can tell theres an issue where a wallet with a trusted node will lose information if the node and wallet where considered synced and the node went into a long sync

there are 2 options for updating the wallet in this case from what i can tell
1. send all the updates to the wallet when we are done -> we dont currently do that when we long sync, we dont aggregrate each batch result and the wallet wont get the updates at the end this https://github.com/Chia-Network/chia-blockchain/blob/782bd01994e03bc58cab646e70677df40141ba5a/chia/full_node/full_node.py#L1482 is where we stage the summary that we send to update wallets inside peak_post_processing_2
2. the wallet will trigger a wallet long sync and ask for everything it needs -> this only happens when the node is marked as not_synced in the wallet but if it was already marked as synced the only way for it to change is a disconnect

still trying to understand whats the better path here, if its aggregating or just telling the wallet to resync

File: chia/_tests/util/misc.py [hunk 1]
```diff
@@ -617,27 +617,6 @@ def marshal(self) -> Dict[str, Any]:
         }
 
 
-T = TypeVar("T")
-
-
-@dataclasses.dataclass(frozen=True)
-class DataTypeProtocol(Protocol):
```
Comment (arvidn) [inline chia/_tests/util/misc.py:624]: I don't know what we use this type for, but it doesn't seem to intersect with the changes you're making. Is this an unrelated clean-up? (if so, it might be good to move it to a separate PR to make it easier to review the main change)

File: chia/consensus/blockchain.py [hunk 2]
```diff
@@ -338,78 +337,18 @@ async def add_block(
...
-                fork_info.reset(block.height - 1, block.prev_header_hash)
-
-            block_rec = await self.get_block_record_from_db(header_hash)
-            if block_rec is not None:
-                # We have already validated the block, but if it's not part of the
-                # main chain, we still need to re-run it to update the additions and
-                # removals in fork_info.
                 await self.advance_fork_info(block, fork_info)
                 fork_info.include_spends(pre_validation_result.conds, block, header_hash)
-                self.add_block_record(block_rec)
```
Comment (arvidn) [inline chia/consensus/blockchain.py:408]: are you sure we don't need this? If this happens because of a deep reorg, `block_rec` may have been pulled from the disk, and may not be in the block cache.
If we're already guaranteed that it's in the cache, can we change the `get_block_record_from_db()` to just `try_block_record()` instead?

File: chia/consensus/blockchain.py [hunk 3]
```diff
@@ -338,78 +337,18 @@ async def add_block(
...
+        block_rec = await self.get_block_record_from_db(header_hash)
+        if block_rec is not None:
             if extending_main_chain:
-                fork_info.reset(block.height - 1, block.prev_header_hash)
-
-            block_rec = await self.get_block_record_from_db(header_hash)
-            if block_rec is not None:
-                # We have already validated the block, but if it's not part of the
-                # main chain, we still need to re-run it to update the additions and
-                # removals in fork_info.
```
Comment (arvidn) [inline chia/consensus/blockchain.py:405]: I think this comment is still relevant and helpful

File: chia/full_node/full_node.py [hunk 4]
```diff
@@ -1306,16 +1303,10 @@ async def add_block_batch(
                     if block_rec.sub_epoch_summary_included.new_difficulty is not None:
                         vs.current_difficulty = block_rec.sub_epoch_summary_included.new_difficulty
 
-            if fork_info is None:
-                continue
             # the below section updates the fork_info object, if
             # there is one.
-
-            # TODO: it seems unnecessary to request overlapping block ranges
-            # when syncing
```
Comment (arvidn) [inline chia/full_node/full_node.py:1315]: maybe you know the answer to this. is it because we can only reorg at the start of a sub-epoch or epoch? So we always overshoot?

**Review by arvidn — COMMENTED** [review]
looks really good. I have a few questions. Was it deliberate to add the `__init__.py` file?

File: chia/_tests/process_junit.py [hunk 5]
```diff
@@ -12,8 +12,8 @@
 import click
 import lxml.etree
 
-from chia._tests.util.misc import BenchmarkData, DataTypeProtocol, TestId
-from chia._tests.util.time_out_assert import TimeOutAssertData
+from chia._tests.util.misc import BenchmarkData, TestId
+from chia._tests.util.time_out_assert import DataTypeProtocol, TimeOutAssertData
```
[the remark this answers is not in the document]
Comment (almogdepaz) [reply chia/_tests/process_junit.py:16]: circular dependency i had to resolve for the changes in misc

File: chia/_tests/util/misc.py (hunk 1 above)
Comment (almogdepaz) [reply chia/_tests/util/misc.py:624]: circular dependency i had to resolve for the changes in misc

File: chia/consensus/blockchain.py (hunk 2 above)
Comment (almogdepaz) [reply chia/consensus/blockchain.py:408]: it looked a littel redundant to me but maybe your right, ill revert this

File: chia/_tests/blockchain/test_blockchain.py [hunk 6]
```diff
@@ -3347,7 +3400,12 @@ async def test_long_reorg(
             assert pre_validation_results[i].error is None
             if (block.height % 100) == 0:
                 print(f"main chain: {block.height:4} weight: {block.weight}")
-            (result, err, _) = await b.add_block(block, pre_validation_results[i], None, sub_slot_iters=ssi)
+
+            fork_info: Optional[ForkInfo] = ForkInfo(block.height - 1, block.height - 1, block.prev_header_hash)
```
[the remark this answers is not in the document]
Comment (almogdepaz) [reply chia/_tests/blockchain/test_blockchain.py:3404]: thanks i originally did all this to avoid using fork_2 which needs to be optional and then opted for just doing new variable per case so its easier to read but i dont mind, changed

File: chia/_tests/util/misc.py [hunk 7]
```diff
@@ -682,22 +648,39 @@ async def add_blocks_in_batches(
...
         assert err is None
         assert success is True
+        if state_change_summary is not None:
+            peak_fb: Optional[FullBlock] = await full_node.blockchain.get_full_peak()
+            assert peak_fb is not None
+            ppp_result: PeakPostProcessingResult = await full_node.peak_post_processing(
+                peak_fb, state_change_summary, dummy_peer
+            )
+            await full_node.peak_post_processing_2(peak_fb, dummy_peer, state_change_summary, ppp_result)
+    await full_node._finish_sync()
```
[the remark this answers is not in the document]
Comment (almogdepaz) [reply chia/_tests/util/misc.py:686]: yes i added the post processing and the finish_sync calls to better emulate the original flow from full_node.py, its needed so we notify the wallets on the peak changes

**Review by arvidn — APPROVED** [review]
This is a great simplification. I think that, perhaps, the original intent shifted a little bit though. The description suggests the fork info isn't necessary, but the change makes it mandatory. i.e. `add_block()` computing fork info isn't necessary.

**Comment by arvidn** [comment]
that missing coverage looks odd.

**Comment by almogdepaz** [comment]
> This is a great simplification. I think that, perhaps, the original intent shifted a little bit though. The description suggests the fork info isn't necessary, but the change makes it mandatory. i.e. `add_block()` computing fork info isn't necessary.

right, i added the change made to add_block_batches to the description
